libmuscle/python/libmuscle/mmsf_validator.py

```diff
@@ -0,0 +1,205 @@
+        # Pretend we're now in F_INIT and we have already received on all F_INIT ports:
+        self._current_operator = Operator.F_INIT
+        self._current_ports_used = self._connected_ports.get(Operator.F_INIT, [])
+
+    def _check_send_receive(
+            self, port_name: str, slot: Optional[int]) -> None:
+        """Actual implementation of check_send/check_receive."""
+        if not self._enabled:
+            return
+
+        operator = self._port_operators[port_name]
+        if self._current_operator != operator:
+            # Operator changed, check that all ports were used in the previous operator
+            self._check_transition(operator, port_name)
+
+        if port_name in self._current_ports_used:
+            # We're using the same port for a second time, this is fine when we're
+            # allowed to do this operator immediately again:
+            self._check_transition(operator, port_name)
+
+        self._current_ports_used.append(port_name)
+
+    def _check_transition(self, operator: Operator, port_name: str = "") -> None:
+        """Check that a transition to the provided operator is allowed.
+
+        Log a warning when the transition does not adhere to the MMSF.
+
+        Args:
+            operator: Operator to transition to.
+            port_name: The name of the port that was sent/receveived on. This is only
+                used for constructing the warning message.
+        """
+        connected_ports = self._connected_ports.get(self._current_operator, [])
+        expected: str = ""
+
+        unused_ports = [
+                port for port in connected_ports
+                if port not in self._current_ports_used]
+        if unused_ports:
+            # We didn't complete the current phase
+            if self._current_operator in (Operator.F_INIT, Operator.S):
+                expected = "a receive"
+            else:
+                expected = "a send"
+            expected += (
+                    f" on any of these {self._current_operator.name} ports: "
+                    + ", ".join(unused_ports))
+
+        elif (self._current_operator, operator) not in self._allowed_transitions:
+            # Transition to the operator is not allowed, now figure out what we were
+            # actually expecting.
+            # First find the allowed transitions from self._current_operator, that are
+            # also 'valid' (i.e. have connected ports):
+            allowed = [
+                    to_op for from_op, to_op in self._allowed_transitions
+                    if from_op is self._current_operator and
+                    (to_op in self._connected_ports or to_op is Operator.NONE)]
+            # Build the message we want to display to users:
+            expected_lst = []
+            for to_op in sorted(allowed, key=lambda op: op.value):
+                ports = ', '.join(map(repr, self._connected_ports.get(to_op, [])))
+                if to_op is Operator.NONE:
```

Comment:
Ah, I see we're using `Operator.NONE` for the stage before the reuse loop. I guess that works as long as the users don't use the operator themselves.